### PR TITLE
Add LinkerFlavor to NativeAOT

### PR DESCRIPTION
**Pull request by am11**
On linux-musl-arm64, for example, the hello world app goes from 2 MB (2049832 B) to 1.8 MB (1918952 B) when `binutils-gold` package is installed.

### CODE REVIEW

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 1]
```diff
@@ -190,6 +190,21 @@ The .NET Foundation licenses this file to you under the MIT license.
     <Error Condition="'$(_WhereLinker)' != '0' and '$(CppCompilerAndLinkerAlternative)' == '' and '$(_IsApplePlatform)' != 'true'"
       Text="Requested linker ('$(CppLinker)') not found in PATH." />
 
+    <![CDATA[
+      Testing if gold linker is available:
+        ! echo "" | cc -xc -fuse-ld=gold - 2>&1 | grep -qE "fuse-ld=gold|'ld'"
+    ]]>
```
Comment (am11) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:196]:
```suggestion
    <!--
      Testing if gold linker is available:
        ! echo "" | cc -xc -fuse-ld=gold - 2>&1 | grep -qE "fuse-ld=gold|'ld'"
    -->
```

**Comment by MichalStrehovsky** [comment]
I pulled down the coredump from one of the failing ARM64 tests:

The stack is:

```
 # Child-SP          RetAddr               Call Site
00 0000ffff`e02dd4d0 0000aae3`2507baf0     System_Runtime_Intrinsics!UnixNativeCodeManager::EHEnumInit+0x8 [/__w/1/s/src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp @ 817] 
01 0000ffff`e02dd4d0 0000aae3`2507b960     System_Runtime_Intrinsics!S_P_CoreLib_System_Runtime_EH::FindFirstPassHandler+0x50 [/_/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs @ 795] 
02 0000ffff`e02dd580 0000aae3`2507b828     System_Runtime_Intrinsics!S_P_CoreLib_System_Runtime_EH::DispatchEx+0xe0 [/_/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs @ 637] 
03 0000ffff`e02dd610 0000aae3`24fbfe00     System_Runtime_Intrinsics!S_P_CoreLib_System_Runtime_EH::RhThrowEx+0x48 [/_/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs @ 573] 
04 0000ffff`e02dd630 0000aae3`25068514     System_Runtime_Intrinsics!RhpThrowEx+0xbc [/__w/1/s/src/coreclr/nativeaot/Runtime/arm64/ExceptionHandling.S @ 338] 
05 0000ffff`e02dd990 0000aae3`250682bc     System_Runtime_Intrinsics!S_P_CoreLib_System_Text_EncodingTable::InternalGetCodePageFromName+0x1f4 [/_/src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs @ 99] 
```

We segfault at this line:

https://github.com/dotnet/runtime/blob/3fec6864a8f728d01f4f4056c025a4aee361ea9e/src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp#L817

I hope the size savings we're seeing are not because gold stripped the LSDA part of the unwinding information :(

**Comment by am11** [comment]
Good and bad news.

Good news is that dwarf-dump errors are lowered with gold:

>         Found 9859 warnings and errors, expected between 12000 and 13000

Bad news: `--gc-section` issue strike again. Just like llvm linker (lld), ld.gold is also eagerly optimizing some sections with `--gc-section` which we depend on. The only difference is lld reports the issues at link time, while gold fails at run time. We probably need some annotations (or [full initialization](https://github.com/libhugetlbfs/libhugetlbfs/issues/49)?) somewhere to help these linkers identify the required sections in the final object and carefully optimize it (currently they drop the sections wholesale).

Stats for hwapp app: release config with stripped symbols

|  Linker flavor | Size in bytes  | With `--gc-sections`  |
|---|---|---|
|  bfd | 2181200  | no  |
|  gold | 2181472 |  no |
| lld  | 2153832 | no  |
| mold | 2150816 | no |
|  bfd | 2049832  | yes  |
|  gold | 1918952 |  yes |
| lld  | _-- fails to link --_ | yes  |
| mold | 1995168 | yes |

I think for now, we can just provide it as an option `-p:LinkerFlavor=` linker with default set to `bfd` on linux and `--gc-sections` disabled for the other flavors; until we figures out a correct way of preserving the section for safe linking.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 2]
```diff
@@ -21,7 +21,8 @@ The .NET Foundation licenses this file to you under the MIT license.
     <CppLibCreator>ar</CppLibCreator>
     <DsymUtilOptions Condition="'$(_IsApplePlatform)' == 'true'">--flat</DsymUtilOptions>
     <_SymbolPrefix Condition="'$(_IsApplePlatform)' == 'true'">_</_SymbolPrefix>
-    <UseLLVMLinker Condition="'$(UseLLVMLinker)' == '' and '$(_targetOS)' == 'freebsd'">true</UseLLVMLinker>
+    <LinkerFlavor Condition="'$(LinkerFlavor)' == '' and '$(_targetOS)' == 'freebsd'">lld</LinkerFlavor>
+    <LinkerFlavor Condition="'$(LinkerFlavor)' == '' and '$(_targetOS)' == 'linux'">gold</LinkerFlavor>
```
Comment (am11) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:25]:
```suggestion
    <LinkerFlavor Condition="'$(LinkerFlavor)' == '' and '$(_targetOS)' == 'linux'">bfd</LinkerFlavor>
```

**Comment by jkotas** [comment]
Just to make sure that I understand what this change does after the latest revision:
- There are no changes in any defaults. We still use the same linker by default as before.
- `UseLLVMLinker` property is replaced by a more flexible `LinkerFlavor` property
Is my understanding correct?

**Comment by am11** [comment]
Yes, this is correct.

If some distro uses `gold`, `mold` or `lld` as default (build time option of gcc and llvm), it will now explicitly select/require `bfd` linker. This is what we want because we use `--gc-sections` with which other linkers strip away some section more eagerly than we expect and cause random runtime failures; hello world works but exceptional paths terminate with wrong signal.

**Review by jkotas — APPROVED** [review]

**Pull request merged**